File: api/api-nestjs/src/signup/signup.service.ts
import { HttpException, HttpStatus, Injectable } from '@nestjs/common';
import { CreateSignupDto } from './dto/create-signup.dto';
import { UpdateSignupDto } from './dto/update-signup.dto';
import * as crypto from 'crypto';
import { SuccessResponse } from '../response/success-response';
import { CustomException } from 'src/exception/custom.exception';

@Injectable()
export class SignupService {
  private usersDb: Record<string, string> = {};

  signup(data: CreateSignupDto): { statusCode: number; content: string } {
    const { id, password } = data;

    if (this.usersDb[id]) {
      throw new CustomException(
        HttpStatus.CONFLICT,
        '이미 사용 중인 ID입니다.'
      );
    }

    const hashedPassword = crypto.createHash('sha256').update(password).digest('hex');

    this.usersDb[id] = hashedPassword;

    const response = new SuccessResponse(HttpStatus.CREATED, '회원가입 성공');
    return response
  }


  create(createSignupDto: CreateSignupDto) {
    return 'This action adds a new signup';
  }

  findAll() {
    return `This action returns all signup`;
  }

  findOne(id: number) {
    return `This action returns a #${id} signup`;
  }

  update(id: number, updateSignupDto: UpdateSignupDto) {
    return `This action updates a #${id} signup`;
  }

  remove(id: number) {
    return `This action removes a #${id} signup`;
  }
}

File: api/api-pynest/src/signup/signup_service.py
from .signup_model import Signup
from nest.core import Injectable
from hashlib import sha256
from typing import Dict
from http import HTTPStatus

class CustomExceptionResponse(Exception):
    def __init__(self, status_code: int, message: str):
        self.status_code = status_code
        self.message = message
        super().__init__(message)

users_db: Dict[str, str] = {}

@Injectable
class SignupService:

    def __init__(self):
        self.database = []
        
    def get_signup(self):
        return self.database
    
    def add_signup(self, signup: Signup):
        id = signup.id
        password = signup.password
        if id in users_db:
            raise CustomExceptionResponse(
                status_code=HTTPStatus.CONFLICT.value,
                message="이미 사용 중인 ID입니다."
            )
            
        hashed_password = sha256(password.encode()).hexdigest()

        users_db[id] = hashed_password

        return "회원가입 성공"
